File: src/python/pcb_to_netlist/5_parse_nets.py
# file: 5_parse_nets.py
# Extract net connections from PcbDoc to nets_raw.txt

import sys # argv, exit

def main():
    if len(sys.argv) != 3: sys.exit(1)
    nets = {}
    with open(sys.argv[1], 'r') as f:
        for l in f:
            if not l.startswith('RECORD=Pad'): continue
            p = {x.split('=')[0]: x.split('=')[1] for x in l.split('|') if '=' in x}
            nid, cid, name = p.get('NET'), p.get('COMPONENT'), p.get('NAME')
            if nid and cid and name: nets.setdefault(nid, []).append(f"{cid}_{name}")
    with open(sys.argv[2], 'w') as f:
        for nid, conns in nets.items(): f.write(f"NET{nid}:{','.join(conns)}\n")

if __name__ == "__main__":
    main()

File: src/typescript/pcb_to_netlist/5_parse_nets.ts
import fs from 'fs';

export function parseNets(inPath: string, outPath: string): void {
    const nets: Record<string, string[]> = {};
    const content = fs.readFileSync(inPath, 'utf8');
    const lines = content.split(/\r?\n/);
    
    for (const line of lines) {
        if (!line.startsWith('RECORD=Pad')) continue;
        const p: Record<string, string> = {};
        line.split('|').forEach(x => {
            if (x.includes('=')) {
                const [k, v] = x.split('=', 2);
                p[k] = v;
            }
        });
        
        const nid = p['NET'];
        const cid = p['COMPONENT'];
        const name = p['NAME'];
        
        if (nid && cid && name) {
            if (!nets[nid]) nets[nid] = [];
            nets[nid].push(`${cid}_${name}`);
        }
    }
    
    const output = Object.entries(nets)
        .map(([nid, conns]) => `NET${nid}:${conns.join(',')}`)
        .join('\n');
    
    fs.writeFileSync(outPath, output);
}
